core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -422,7 +430,7 @@ class BlockManagerMasterEndpoint(
     val locations = blockLocations.get(blockId)
     if (locations != null) {
       locations.foreach { blockManagerId: BlockManagerId =>
-        val blockManager = blockManagerInfo.get(blockManagerId)
+        val blockManager = blockManagerInfo.get(blockManagerId).filter(_.isAlive)
```

Comment:
Hmm...uses `isActive` is kind of messy. What if we maintain the inactive BlockManagerInfos in a separate data structure and remove the BlockManagerInfo from the `blockManagerInfo` as it is.

(I have an impression we discussed this but I can't remember why we abandon it. Sorry if I raise the duplicate discussion here.)

Reply:
> What if we maintain the inactive BlockManagerInfos in a separate data structure and remove the BlockManagerInfo from the blockManagerInfo as it is.

Yes @Ngone51, you are right, we discussed this earlier. My original solution was exactly what you described, to have a Guava cache which stored inactive BlockManagerInfos. In this case, we also had to pass `InactiveBlockManagerCache` to `BlockManagerMasterHeartbeatEndpoint` so it can respond appropriately to `BlockManagerHeartbeat`.

Later @attilapiros suggested to model BlockManager removal as a new state by adding `removalTS` to `BlockManagerInfo`. I found his solution was better than using the `InactiveBlockManagerCache`.

Reply:
What about introducing some helper methods? 

Sth like:

```scala
 private def aliveBlockManagerInfo(bmId: BlockManagerId): Option[BlockManagerInfo] =
     blockManagerInfo.get(bmId).filter(_.isAlive)

 private def allAliveBlockManagerInfos() =
     blockManagerInfo.values.filter(_.isAlive)
```

They even could be inline methods...

@Ngone51  WDYT?

Reply:
I think the helper methods don't solve the problem thoroughly as you still have to replace all the usages where `isActive` exists now.

I'm personally ok with pass the inactive BlockManagerInfos to `BlockManagerMasterHeartbeatEndpoint`.

@mridulm what's your opinion?

Reply:
I am slightly more in favor of @attilapiros's suggestion .. I am more concerned with future evolution, where someone might accidentally forget to check for `isAlive` : the helper methods should defend against that.

I am fine with either approach though - in current state and with this enhancement, I am fine with this going through.

Reply:
I will add the helper methods as they would clearly help to prevent accidental misses for `isAlive`.

Reply:
Sure, let's go with the way of helper methods.

Reply:
Done.